integration/helpers.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+func parseOutput[T any](output string, normalize func(*T)) ([]*T, error) {
+	ret := []*T{}
+
+	scanner := bufio.NewScanner(strings.NewReader(output))
+	for scanner.Scan() {
+		var entry T
+
+		line := scanner.Text()
+		if err := json.Unmarshal([]byte(line), &entry); err != nil {
+			return nil, fmt.Errorf("unmarshaling line %q: %w", line, err)
+		}
+
+		// To be able to use reflect.DeepEqual and cmp.Diff, we need to
+		// "normalize" the output so that it only includes non-default values
+		// for the fields we are able to verify.
+		if normalize != nil {
+			normalize(&entry)
+		}
+
+		ret = append(ret, &entry)
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("parsing output: %w", err)
+	}
+
+	return ret, nil
+}
+
+// expectAllToMatch verifies that all the entries in the output matches the
+// expected entry.
+func expectAllToMatch[T any](output string, expectedEntry *T, normalize func(*T)) error {
+	entries, err := parseOutput(output, normalize)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		if !reflect.DeepEqual(expectedEntry, entry) {
+			return fmt.Errorf("unexpected output entry:\n%s\n",
+				cmp.Diff(expectedEntry, entry))
+		}
+	}
+
+	return nil
+}
+
+// expectAtLeastOneToMatch verifies that at least one entry in the output
+// matches the expected entry.
+func expectAtLeastOneToMatch[T any](output string, expectedEntry *T, normalize func(*T)) error {
+	entries, err := parseOutput(output, normalize)
+	if err != nil {
+		return err
+	}
+
+	for _, entry := range entries {
+		if reflect.DeepEqual(expectedEntry, entry) {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("output doesn't contain the expected entry: %+v", expectedEntry)
+}
+
+// expectListIsMatched verifies that all the entries in expectedEntries is
+// matched by at least one entry in the output.
+func expectListIsMatched[T any](output string, expectedEntries []*T, normalize func(*T)) error {
```

Comment:
Can we combine `expectAtLeastOneToMatch` and `expectListIsMatched` into a single function: `expectEventsToMatch(output string, normalize func(*T), expectedEntries ...*T)`?

Reply:
Agree! I think we are converging on a solution:

```golang
// expectAllToMatch verifies that the expectedEntry is matched by all the
// entries in the output.
func expectAllToMatch[T any](output string, normalize func(*T), expectedEntry *T) error
```

```golang
// expectAtLeastOneToMatch verifies that all the elements in expectedEntries are
// matched by at least one entry in the output.
func expectAtLeastOneToMatch[T any](output string, normalize func(*T), expectedEntries ...*T)
```

Regarding `expectAllToMatch` (or `expectAllEntriesToMatch`), if we expect all the entries to match, it is obvious that it receives only one `expectedEntry`, so it is fine. Instead, in the case of `expectAtLeastOneToMatch` (or `expectAtLeastOneEntryToMatch`), we are more flexible, and it could receives multiple elements that must be matched by at least one entry in the output. WDYT?

Reply:
I think `One` in `expectAtLeastOneEntryToMatch` is a bit confusing when multiple entries are passed. I'd go for a more generic name like `expectEntriesToMatch`. (Also, if we really want we can provide a `expectAtLeastOneToMatch` wrapper that only accepts a single entry).

Reply:
> I think `One` in `expectAtLeastOneEntryToMatch` is a bit confusing when multiple entries are passed. 
Good point.